mongo-store.js

```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
     },
 
     save: function (args, cb) {
-      var ent = args.ent
+      const is_update = Boolean(args.ent.id)
 
-      var update = !!ent.id
+      if (is_update) {
+        return updateExisting(args, cb)
+      }
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+      return createAndSave(args, cb)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+      function createAndSave(args, cb) {
+        return upsertIfRequested(args, function (err, upsert) {
+          if (error(args, err, cb)) {
+            return
           }
 
-          entp._id = makeid(id)
+          if (upsert.upsert_requested) {
+            return cb(null, upsert.out)
+          }
+
+          return createNew(args, cb)
+        })
+
+
+        function upsertIfRequested(args, cb) {
+          const query_for_save = args.q
+
+          if (Array.isArray(query_for_save.upsert$)) {
```

Comment:
convert to boolean, perhaps

`const has_upsert = Array.isArray(query_for_save.upsert$)`